File: output/python/tests/21_actions_basic.py
from typing import Any, Optional, List, Dict, Callable

class ActionsTestFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class ActionsTestFrameContext:
    def __init__(self, event: ActionsTestFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class ActionsTestCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'ActionsTestCompartment':
        c = ActionsTestCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class ActionsTest:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.log = ""
        self.__compartment = ActionsTestCompartment("Ready")
        self.__next_compartment = None
        __frame_event = ActionsTestFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = ActionsTestFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = ActionsTestFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = ActionsTestFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def process(self, value: int) -> int:
        __e = ActionsTestFrameEvent("process", {"value": value})
        __ctx = ActionsTestFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def get_log(self) -> str:
        __e = ActionsTestFrameEvent("get_log", None)
        __ctx = ActionsTestFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def _state_Ready(self, __e):
        if __e._message == "get_log":
            self._context_stack[-1]._return = self.log
            return
        elif __e._message == "process":
            value = __e._parameters["value"]
            self.__log_event("start")
            self.__validate_positive(value)
            self.__log_event("valid")
            result = value * 2
            self.__log_event("done")
            self._context_stack[-1]._return = result
            return

    def __log_event(self, msg: str):
        self.log = self.log + msg + ";"

    def __validate_positive(self, n: int):
        if n < 0:
            raise ValueError(f"Value must be positive: {n}")


def main():
    print("=== Test 21: Actions Basic ===")
    s = ActionsTest()

    # Test 1: Actions are called correctly
    result = s.process(5)
    assert result == 10, f"Expected 10, got {result}"
    print(f"1. process(5) = {result}")

    # Test 2: Log shows action calls
    log = s.get_log()
    assert "start" in log, f"Missing 'start' in log: {log}"
    assert "valid" in log, f"Missing 'valid' in log: {log}"
    assert "done" in log, f"Missing 'done' in log: {log}"
    print(f"2. Log: {log}")

    # Test 3: Action with validation
    try:
        s.process(-1)
        assert False, "Should have raised ValueError"
    except ValueError as e:
        print(f"3. Validation caught: {e}")

    print("PASS: Actions basic works correctly")

if __name__ == '__main__':
    main()

File: output/typescript/tests/21_actions_basic.ts
class ActionsTestFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
    }
}


class ActionsTestFrameContext {
    public event: ActionsTestFrameEvent;
    public _return: any;
    public _data: Record<string, any>;

    constructor(event: ActionsTestFrameEvent, default_return: any) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


class ActionsTestCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: ActionsTestCompartment | null;

    constructor(state: string, parent_compartment: ActionsTestCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): ActionsTestCompartment {
        const c = new ActionsTestCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class ActionsTest {
    private _state_stack: Array<any>;
    private __compartment: ActionsTestCompartment;
    private __next_compartment: ActionsTestCompartment | null;
    private _context_stack: Array<any>;
    private log: string = "";

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.log = "";
        this.__compartment = new ActionsTestCompartment("Ready");
        this.__next_compartment = null;
        const __frame_event = new ActionsTestFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: ActionsTestFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new ActionsTestFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new ActionsTestFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new ActionsTestFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: ActionsTestFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: ActionsTestCompartment) {
        this.__next_compartment = next_compartment;
    }

    public process(value: number): number {
        const __e = new ActionsTestFrameEvent("process", {"value": value});
        const __ctx = new ActionsTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    public get_log(): string {
        const __e = new ActionsTestFrameEvent("get_log", null);
        const __ctx = new ActionsTestFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()!._return;
    }

    private _state_Ready(__e: ActionsTestFrameEvent) {
        if (__e._message === "get_log") {
            this._context_stack[this._context_stack.length - 1]._return = this.log;
            return;;
        } else if (__e._message === "process") {
            const value = __e._parameters?.["value"];
            this.__log_event("start");
            this.__validate_positive(value);
            this.__log_event("valid");
            const result = value * 2;
            this.__log_event("done");
            this._context_stack[this._context_stack.length - 1]._return = result;
            return;;
        }
    }

    private __log_event(msg: string) {
                    this.log = this.log + msg + ";";
    }

    private __validate_positive(n: number) {
                    if (n < 0) {
                        throw new Error(`Value must be positive: ${n}`);
                    }
    }
}


function main(): void {
    console.log("=== Test 21: Actions Basic (TypeScript) ===");
    const s = new ActionsTest();

    // Test 1: Actions are called correctly
    const result = s.process(5);
    if (result !== 10) {
        throw new Error(`Expected 10, got ${result}`);
    }
    console.log(`1. process(5) = ${result}`);

    // Test 2: Log shows action calls
    const log = s.get_log();
    if (!log.includes("start")) {
        throw new Error(`Missing 'start' in log: ${log}`);
    }
    if (!log.includes("valid")) {
        throw new Error(`Missing 'valid' in log: ${log}`);
    }
    if (!log.includes("done")) {
        throw new Error(`Missing 'done' in log: ${log}`);
    }
    console.log(`2. Log: ${log}`);

    // Test 3: Action with validation
    try {
        s.process(-1);
        throw new Error("Should have thrown Error");
    } catch (e) {
        if (e instanceof Error && e.message.includes("positive")) {
            console.log(`3. Validation caught: ${e.message}`);
        } else {
            throw e;
        }
    }

    console.log("PASS: Actions basic works correctly");
}

main();
